tango feedstocks submission to conda-forge preparation

beenje opened the pull request:
This is a local PR to prepare submission to conda-forge.
This is based on the existing recipes from https://github.com/tango-controls/tango-conda-recipes and https://github.com/tango-controls/pytango-conda-recipes with some changes following the template from conda-forge.

@bourtemb, I added you as maintainer to the following recipes:
 - tango-idl
 - cpptango
 - tango-test
 - tango-admin

Is this OK? Shall I add someone else?
Could you please review those recipes? I bumped the build number compared to the existing recipes in tango channel. Not sure if we should or not.

@ajoubertza, I added you as maintainer to:
- pytango
- itango

Is this OK? Someone else to add?
For pytango, I removed the `pytest` command. In most recipes, the full test suite isn't run. But that's really up to the maintainer to decide. What do you think? Should we add it back?

bourtemb commented:
@beenje , thank you very much for what you're doing!

You can add me in the list of maintainers for the packages you listed for me, even though I haven't done anything on the tango-admin one yet. 
If we could find some other volunteers, that would be great. Thank you for adding yourself in the list of maintainers.
Help is always welcome.

I would be in favour of keeping the pytest part for pytango.
I think it's good practice to test the packages, isn't it? Or would this be considered as too much and too slow?

bourtemb commented on recipes/cpptango/build.sh:
```diff
@@ -0,0 +1,16 @@
+mkdir build
+cd build
+cmake -DCMAKE_BUILD_TYPE=Relase \
```
There is a typo here. Relase => Release

I see you're trying to change from Debug to Release mode.
Please be aware that this means that it will be difficult to debug if there is an issue.
I think it makes sense to provide a Release version for better performance but we should probably find a way for the people to easily install a debug version of the package...
I don't know how other C++ library conda packages deal with that.

beenje replied:
Thanks for the typo!
I thought it makes more sense to build in release mode when creating a package. This is what is done for most Conda packages as far as I know.
After I guess it depends what the usage will be. What could maybe be possible is to create 2 outputs: `cpptango` and `cpptango-debug`. To build in debug mode, do we need to have the dependent libraries available in debug as well or not? On windows I noticed that in debug mode, it's looking for debug libraries. They won't be available with Conda.

ajoubertza replied:
For PyTango debugging, testing and development (which is done via Conda on Linux), it is very useful to have a debug version of cppTango.  I'm not sure if any sites are planning to do production deployments via Conda.  Production deployments may benefit from the increased performance from a release mode build.

If we can only have one package, then it should be debug (at least for Linux).  If we can have multiple (and it is easy to maintain), then adding `cpptango-debug` would be OK.

beenje replied:
After looking a bit more, it might not be that simple. Just creating 2 cppTango packages would make it a bit difficult for the other packages (we would need one pytango-debug depending on cpptango-debug...). From what I read, it could be achieved with track features.
Anyway, it's probably not worth the effort for now.
I switched to `Debug` type. We can always investigate later if we do want a `Release` one.

bourtemb commented on recipes/cpptango/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+
+about:
+  home: https://www.tango-controls.org
+  license: LGPL-3.0-only
```
```suggestion
  license: LGPL-3.0-or-later
```

bourtemb commented on recipes/itango/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+
+about:
+  home: https://github.com/tango-controls/itango
+  license: LGPL-3.0-only
```
```suggestion
  license: LGPL-3.0-or-later
```

beenje commented:
Thanks @bourtemb!
You are probably right about `pytango`. I added back the `pytest` command.
More testing is always good. It's actually not that long to run, so there is no good reason to skip it.

Yes, it would be nice to have more volunteers. Something to ask during next tango meeting?

Do you have an opinion about the build number? Should we just reset it to 0 as it's for a new channel? Or should we try to increase the number compared to the existing packages in `tango-controls`?

bourtemb commented on recipes/pytango/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+
+about:
+  home: http://pytango.rtfd.io
+  license: LGPL-3.0-only
```
```suggestion
  license: LGPL-3.0-or-later
```

bourtemb commented on recipes/tango-admin/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+
+about:
+  home: https://www.tango-controls.org
+  license: GPL-3.0-only
```
```suggestion
  license: GPL-3.0-or-later
```

bourtemb commented on recipes/tango-admin/tango_admin-install.patch:
```diff
@@ -0,0 +1,10 @@
+ target_link_libraries(${PROJECT_NAME} PUBLIC ${TANGO_PKG_LIBRARIES} ${CMAKE_DL_LIBS})
+ 
+-#TODO install
++install(TARGETS ${PROJECT_NAME} RUNTIME DESTINATION bin)
```
I think the best would be to add it in the tango_admin repository.

beenje replied:
I already made a PR that was merged: https://github.com/tango-controls/tango_admin/pull/4
But to submit this version, the patch required. We can remove it in next release.

bourtemb replied:
I understand. 
Then I guess, we should simply make a new release 1.16 of tango_admin?

bourtemb replied:
tango_admin Release_1.16 is out and also includes a LICENSE file now.

beenje replied:
Cool. Thanks! Will update the recipe.

beenje replied:
tango-admin recipe updated to 1.16

bourtemb commented on recipes/tango-idl/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+about:
+about:
+  home: http://www.tango-controls.org
+  license: LGPL-3.0-only
```
Not sure about the appropriate license for tango-idl.
If we go for LGPL-3.0, I think we should use LGPL-3.0-or-later

beenje replied:
I opened a PR: https://github.com/tango-controls/tango-idl/pull/11
And changed to `LGPL-3.0-or-later`.

beenje replied:
I bumped tango-idl to 5.1.0

bourtemb commented on recipes/tango-test/build.sh:
```diff
@@ -0,0 +1,9 @@
+mkdir build
+cd build
+cmake -DCMAKE_BUILD_TYPE=Release \
```
I would keep TangoTest in Debug mode because it is used mainly to test things and debugging issues.
Of course, this would not be optimal in the case of benchmarks.

beenje replied:
As for `cppTango`, I switched to `Debug`.

bourtemb commented on recipes/tango-test/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+
+about:
+  home: https://www.tango-controls.org
+  license: GPL-3.0-only
```
```suggestion
  license: GPL-3.0-or-later
```

bourtemb reviewed (commented):
Thanks @beenje for this great initiative.

I made a few comments on the Debug/Release mode and on appropriate licenses.

I think the licenses should be added to the upstream repositories (as already done for some of them like pytango) and not added here.

tango_admin install patch should rather be in tango_admin upstream repo.

bourtemb commented:
> Do you have an opinion about the build number? Should we just reset it to 0 as it's for a new channel? Or should we try to increase the number compared to the existing packages in `tango-controls`?

The safest approach is probably to increase the build number. I don't know how conda deals with this when you have the same package in different channels.

beenje commented:
I'll make MR for missing LICENSE upstream. The files can stay in the recipe depending if we can make a new release or not.

ajoubertza commented:
> @ajoubertza, I added you as maintainer to:
> 
> * pytango
> * itango
> 
> Is this OK? Someone else to add?

Thanks.  I'm not sure of anyone else to add.  Maybe @cpascual is interested?

> For pytango, I removed the `pytest` command. In most recipes, the full test suite isn't run. But that's really up to the maintainer to decide. What do you think? Should we add it back?

I would like to keep the tests to ensure that the Conda build is working, as @bourtemb suggests.   At least for Linux.

This PR doesn't cover Windows, but for future reference:  the Windows the testing is more complicated and slower, so I'm OK with it being skipped - it gets done on AppVeyor, and we install binary wheels for Windows, so no additional compilation required.

beenje commented:
> > Do you have an opinion about the build number? Should we just reset it to 0 as it's for a new channel? Or should we try to increase the number compared to the existing packages in `tango-controls`?
> 
> The safest approach is probably to increase the build number. I don't know how conda deals with this when you have the same package in different channels.

I know that by default if a package is available in different channels, the one from the first defined channel (it has highest priority) will be used. I wasn't sure when build numbers are different, but I did a test. Even if `foo_1` is available in channel A and `foo_0` in channel B, what matters if the order of the channel.
So actually I think we should reset the build number to 0 for all packages.

bourtemb commented:
@beenje I created release 3.2 for TangoTest (https://github.com/tango-controls/TangoTest/releases/tag/3.2) which is including the LICENSE file.
For itango, we would need a new release too if possible.
@ajoubertza or another itango maintainer, could you make a new itango release, please, if it is in a reasonable state for a new release?
Latest release itango v0.1.7 is from October 2017. There has been new commits since that time.

bourtemb commented:
For tango-idl, I'm not sure what is the best license we should use.
Please comment https://github.com/tango-controls/tango-idl/pull/11 if you have a clearer idea than me.
Maybe @cpascual, @mliszcz or @t-b would have a better idea?

mliszcz commented:
I can't help much here because I am not well versed in law or licensing. But you may want to read this:
https://opensource.stackexchange.com/questions/6058/is-code-generated-from-an-interface-definition-file-a-derivative-work
Per my understanding, if I generate any code from the Tango IDL file, this code retains the license of the IDL file. This may be relevant in case e.g. one makes a proprietary application that talks to Tango, as one will now include GPL/LGPL code even if they won't link their application with cppTango.

t-b commented:
LGPL for tango-idl is fine with me. Normal GPL IMHO will not work.

beenje commented:
> @beenje I created release 3.2 for TangoTest (https://github.com/tango-controls/TangoTest/releases/tag/3.2) which is including the LICENSE file.

Thanks. I bumped the conda recipe to 3.2.

> For itango, we would need a new release too if possible.
> @ajoubertza or another itango maintainer, could you make a new itango release, please, if it is in a reasonable state for a new release?
> Latest release itango v0.1.7 is from October 2017. There has been new commits since that time.

For itango, I don't think we need a new release for the conda recipe. It's fine to have the LICENSE in the recipe as long as it has been submitted upstream.
After I agree that it would be good to make a new release of itango. I can look into it but would like to migrate to GitLab first so it can be done via CI.

bourtemb commented:
I made a new release (5.1.0) of tango-idl which also includes a LICENSE file.
LGPLv3 was used as originally suggested by @beenje in his PR and following comments on the current PR.

bourtemb commented on recipes/pytango/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - python
+    - pip
+    - cppzmq
+    - cpptango >=9.3,<9.4
```
We might get some issues when we will be releasing some 9.4 pre-release versions because they will be considered as <9.4 
For instance, I have such pre-release test versions on my Reyn4ld conda channel (https://anaconda.org/Reyn4ld/cpptango/files).
I don't know what would be the correct upper value we should put to say < any 9.4 version, including the release candidates, alpha and beta versions.

beenje replied:
Good catch. It should be `<9.4.0a0` to exclude alpha, beta and rc.

bourtemb commented on recipes/pytango/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - cppzmq
+    - cpptango >=9.3,<9.4
+    - boost
+    - numpy
```
The original pytango conda recipe was specifying numpy >=1.1
Shouldn't we keep that for the people who will be manually building the package and would like to install by mistake pytango on an environment where a numpy < 1.1 was installed?

beenje replied:
I was only thinking about conda-forge when updating the recipe. I know that numpy is pinned to 1.16.
Not sure who could build a package with a version older than 1.1 (it's from 2008). But it doesn't cost much to leave it and I guess it gives some useful information. I'll add it back.

bourtemb commented on recipes/tango-idl/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - if not exist %LIBRARY_INC%\tango.idl exit 1 .           # [win]
+    - if not exist %LIBRARY_LIB%\pkgconfig\tangoidl.pc exit 1 # [win]
+
+about:
+about:
```
Duplicated line

beenje replied:
Thanks!

beenje commented:
@bourtemb and @ajoubertza I think I addressed all your comments. Do you agree or have something else to mention?

If you think the current recipes are in a good enough state, I propose to make a PR to conda-forge. I might rebase to keep one commit per recipe before that.

bourtemb reviewed (commented):
@beenje , thanks again for this initiative.
This PR seems in a good shape to me.
I think you could start making a PR to conda-forge based on this work.
Rebasing to keep one commit per recipe before that seems like a good idea indeed.
